projet/phase4/HK.jl

```diff
@@ -0,0 +1,118 @@
+"""Implémente l'algorithme de Held & Karp sur un graphe."""
+function HK(graphe_input::AbstractGraph{T}, iteration::Int, index_racine::Int) where T
+    # Initialisation des variables nécessaires
+    # graphe initial dont le poids des arêtes va être modifié
+    graphe = deepcopy(graphe_input)
+    nodes_graphe = getNodes(graphe)
+    edges_graphe = getEdges(graphe)
+    edges_true_weight = deepcopy(edges_graphe)
+
+    mst = initGraphPrim(graphe)
+    # données pour les itération de HK
+    n = length(nodes_graphe)
+    pi = zeros(n)
+    d = ones(n)
+    t = 1
+    k = 0
+    W = 0
+    sum = 0
+
+
+
+    # Construction du tableau de noeud et d'arêtes pour la construction de l'arbre de recouvrement minimal
+    nodes_mst = copy(nodes_graphe)
+    racine = nodes_mst[index_racine]
+    deleteat!(nodes_mst, index_racine)
+
+    edges_mst = copy(edges_graphe)
+    index_edges = findall(x -> isequal(x.node1, racine) || isequal(x.node2, racine), edges_mst)
+    edges_racine = edges_mst[index_edges]
+    deleteat!(edges_mst, index_edges)
+
+    edge1, edge2 = edges_min!(edges_racine)
```

Comment:
Les deux arêtes reliant la racine au 1-arbre dépendent du poids _avec modification_ : ces deux arêtes doivent être re-calculées à chaque itération !